### ISSUE TITLE
Fails to build on latest dev branch of THREE.js

### ISSUE THREAD

**Issue by coballast** [coballast/threejs-browserify-conversion-utility#10]
Right now, running a browserify build off of the latest dev branch crashes. It works fine with r70 master.

At first I thought I made a change that broke something. If I checkout dev, do

```
node browerifyify ../three.js
gulp build
```
I get

```
[04:52:29] Using gulpfile ~/threejs-browserify-conversion-utility/gulpfile.js
[04:52:29] Starting 'build'...
[04:52:30] Finished 'build' after 34 ms
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: EMFILE, open '/home/colin/threejs-browserify-conversion-utility/package.json'
    at Error (native)
```

But if I checkout r70, everything works fine.

**Comment by coballast** [comment]
This might either be a bug with the utility, or a bug in the dev branch of threejs, or a bug in browserify. I am inclined to think threejs is the problem here.

**Comment by kumavis** [comment]
I got this error when using [browserify-graph](https://github.com/conradz/browserify-graph ) on our converted three.js (master). My guess is more circular dep problems. If circular deps work in node but not in browserify, maybe browserify needs a patch?

**Comment by kumavis** [comment]
You can try increasing the number of files that can be opened, but I feel like thats not the real problem since its stuck on the root package.json.....

**Comment by kumavis** [comment]
Here is a circular deps issue on browserify https://github.com/substack/node-browserify/issues/1068
No insight on resolution -- I think browserify is a bit of a black box for a lot of people

**Comment by kumavis** [comment]
[Here I played with circular deps](https://github.com/kumavis/browserify-circle)... take a look at `late/` and how that resolves some problems...

maybe all exports can be functions and then when we define them on THREE we call the functions... too weird? Do we end up with classes that don't `instanceof` correctly? (We shouldnt be using `instanceof` any ways, but thats a different issue.)

**Comment by kumavis** [comment]
More on circular deps -- I added a new pattern `early/` that [works pretty well](https://github.com/kumavis/browserify-circle/blob/master/early/a.js). The point is to get your reference into `module.exports` before calling require on anything else. For Classes, this will be easy b/c of function hoisting.

**Comment by kumavis** [comment]
I'm getting:
```
╭─{🙈 } kumavis in ~/Development/Node/threejs-browserify-conversion-utility on (master) 
╰─± gulp
[12:47:46] Using gulpfile ~/Development/Node/threejs-browserify-conversion-utility/gulpfile.js
[12:47:46] Starting 'default'...
[12:47:46] Starting 'build'...
[12:47:46] Finished 'build' after 20 ms
[12:47:46] Finished 'default' after 21 ms

events.js:72
        throw er; // Unhandled 'error' event
              ^
Error: EMFILE, open '/Users/kumavis/Development/Node/threejs-browserify-conversion-utility/package.json'
```

on `three.js#master` now

**Comment by kumavis** [comment]
weird -- using a gulputil logger displays a different path that its erroring on:
```
Error: EMFILE, open '/Users/kumavis/Development/Node/threejs-browserify-conversion-utility/src/package.json'
```

notice the `/src/package.json'` not going to find a package.json there! no idea how those paths are being confused or why the errors don't match.


```js
var gutil      = require('gulp-util')

  // ....

  .bundle()

  .on('error', gutil.log.bind(gutil, 'Browserify Error'))
```

**Comment by kumavis** [comment]
idk what all that noise is in the last comment, but its got to be another manifestation of the circular dep error.

**Comment by coballast** [comment]
The problem is definitely with browserify or the three.js source code and not with gulp. If I call browserify manually like this:
```
~/threejs-browserify-conversion-utility$ browserify ./src/Three.js > build/three.min.js
```

I get the exact same error.

**Comment by kumavis** [comment]
@coballast  can you generate a dep graph or do you get the same problem?

**Comment by coballast** [comment]
@kumavis I don't believe there is anything wrong in principle with any of the utility code that generates the browserify three source, or the dependency graphs. Those seem to work fine, no problem. The issue only comes when browserify actually runs.

**Comment by coballast** [comment]
@kumavis I have a branch up with the browserified version of the Three.js source:
https://github.com/coballast/three.js/tree/browserify
I've made some manual corrections to fix problems. I intend to keep it continuously updated with changes from dev using the utility.

The problem is still present there unfortunately, which leads me to think that the trouble is with the source somehow. It is now trying to read the root package.json:
```
colin@colin-CM5671:~/three.js$ gulp
[13:24:33] Using gulpfile ~/three.js/gulpfile.js
[13:24:33] Starting 'default'...
[13:24:33] Starting 'build'...
[13:24:33] Finished 'build' after 29 ms
[13:24:33] Finished 'default' after 30 ms
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: EMFILE, open '/home/colin/three.js/package.json'
    at Error (native)
```

**Comment by coballast** [comment]
The gulp now produces a three.min.js successfully (at least on my machine). This change occurred after I fixed #15 My theory is the excessive and unnecessary imports were giving browserify trouble. Someone double check to make sure this dragon is finally slain, but closing this for now.

**Comment by kumavis** [comment]
yeah i think there must be some exotic case of circular deps it doesnt handle well
nice to see this issue closed.
